File: app-service-comments/src/main/java/com/github/arhor/aws/graphql/federation/comments/data/entity/UserRepresentation.java
package com.github.arhor.aws.graphql.federation.comments.data.entity;

import com.github.arhor.aws.graphql.federation.starter.core.data.Features;
import jakarta.annotation.Nonnull;
import lombok.Builder;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.Immutable;
import org.springframework.data.annotation.PersistenceCreator;
import org.springframework.data.annotation.Transient;
import org.springframework.data.domain.Persistable;
import org.springframework.data.relational.core.mapping.Column;
import org.springframework.data.relational.core.mapping.Table;

import java.util.EnumSet;
import java.util.UUID;

@Table("user_representations")
@Immutable
@Builder(toBuilder = true)
public record UserRepresentation(
    @Id
    @Column("id")
    UUID id,

    @Column("features")
    UserFeatures features,

    @Transient
    boolean shouldBePersisted
) implements Persistable<UUID>, Commentable<UserRepresentation> {

    public UserRepresentation {
        if (features == null) {
            features = new UserFeatures();
        }
    }

    @PersistenceCreator
    public UserRepresentation(final UUID id, final UserFeatures features) {
        this(id, features, false);
    }

    @Override
    public UUID getId() {
        return id;
    }

    @Override
    public boolean isNew() {
        return shouldBePersisted;
    }

    @Nonnull
    @Override
    public UserRepresentation toggleComments() {
        return toBuilder()
            .features(features.toggle(UserFeature.COMMENTS_DISABLED))
            .build();
    }

    @Override
    public boolean commentsDisabled() {
        return features().check(UserFeature.COMMENTS_DISABLED);
    }

    public enum UserFeature {
        COMMENTS_DISABLED,
    }

    public static final class UserFeatures extends Features<UserFeatures, UserFeature> {

        public UserFeatures() {
            super(EnumSet.noneOf(UserFeature.class));
        }

        public UserFeatures(@Nonnull final EnumSet<UserFeature> items) {
            super(EnumSet.copyOf(items));
        }

        @Nonnull
        @Override
        protected UserFeatures create(@Nonnull final EnumSet<UserFeature> items) {
            return new UserFeatures(items);
        }
    }
}

File: app-service-posts/src/main/kotlin/com/github/arhor/aws/graphql/federation/posts/data/entity/UserRepresentation.kt
package com.github.arhor.aws.graphql.federation.posts.data.entity

import com.github.arhor.aws.graphql.federation.starter.core.data.Features
import org.springframework.data.annotation.Id
import org.springframework.data.annotation.Immutable
import org.springframework.data.annotation.PersistenceCreator
import org.springframework.data.annotation.Transient
import org.springframework.data.domain.Persistable
import org.springframework.data.relational.core.mapping.Column
import org.springframework.data.relational.core.mapping.Table
import java.util.EnumSet
import java.util.UUID

@Table(UserRepresentation.TABLE_NAME)
@Immutable
data class UserRepresentation @PersistenceCreator constructor(
    @Id
    @Column("id")
    private val id: UUID,

    @Column("features")
    val features: UserFeatures = UserFeatures(),

    @Transient
    val shouldBePersisted: Boolean = false,
) : Persistable<UUID> {

    override fun getId(): UUID = id

    override fun isNew(): Boolean = shouldBePersisted

    fun togglePosts(): UserRepresentation {
        return copy(features = features.toggle(UserFeature.POSTS_DISABLED))
    }

    fun postsDisabled(): Boolean {
        return features.check(UserFeature.POSTS_DISABLED);
    }

    enum class UserFeature {
        POSTS_DISABLED,
    }

    class UserFeatures(items: EnumSet<UserFeature>) : Features<UserFeatures, UserFeature>(items) {
        constructor()
            : this(EnumSet.noneOf(UserFeature::class.java))

        constructor(feature: UserFeature, vararg features: UserFeature)
            : this(EnumSet.of(feature, *features))

        override fun create(items: EnumSet<UserFeature>) = UserFeatures(items)
    }

    companion object {
        const val TABLE_NAME = "user_representations"
    }
}
